### PR TITLE
Add jose list command

### PR DESCRIPTION
**Pull request by tiran**
The new cmd/jose list commands lists all registered types, ops,
generators etc.. It's main purpose is introspection and debugging
problems like #13.

```
$ cmd/jose list
jwk_type: EC, RSA, oct
jwk_op: [enc]wrapKey/unwrapKey, [enc]encrypt/decrypt, [sig]verify/sign
jwk_generator: EC, RSA, oct
jwk_hasher: sha512, sha384, sha256, sha224, sha1
jws_signer: RS256, RS384, RS512, PS256, PS384, PS512, ES256, ES384, ES512, HS256, HS384, HS512
jwe_crypter: A128GCM, A192GCM, A256GCM, A128CBC-HS256, A192CBC-HS384, A256CBC-HS512
jwe_wrapper: A128KW, A192KW, A256KW, PBES2-HS256+A128KW, PBES2-HS384+A192KW, PBES2-HS512+A256KW, RSA1_5, RSA-OAEP, RSA-OAEP-256,
A128GCMKW, A192GCMKW, A256GCMKW, ECDH-ES, ECDH-ES+A128KW, ECDH-ES+A192KW, ECDH-ES+A256KW, dir
jwe_zipper: DEF
```

Signed-off-by: Christian Heimes <christian@python.org>

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -59,6 +59,7 @@ cmd_jose_SOURCES = \
     cmd/ver.c \
     cmd/enc.c \
     cmd/dec.c \
+    cmd/list.c \
```
Comment (npmccallum) [inline Makefile.am]: Stylistically, all commands are three letters.

File: cmd/jose.h [hunk 2]
```diff
@@ -82,3 +85,6 @@ jcmd_enc(int argc, char *argv[]);
 
 int
 jcmd_dec(int argc, char *argv[]);
+
+int
+jcmd_list(int argc, char *argv[]);
```
Comment (npmccallum) [inline cmd/jose.h]: Need a newline here.

File: cmd/list.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    jose_jwk_op_t *op;
+    jose_jwk_generator_t *generator;
+    jose_jwk_hasher_t *hasher;
+    jose_jws_signer_t *signer;
+    jose_jwe_crypter_t *crypter;
+    jose_jwe_wrapper_t *wrapper;
+    jose_jwe_zipper_t *zipper;
+    char **p;
+
+    for (int c; (c = getopt_long(argc, argv, "hk:i:o:cdt:s:", opts, NULL)) >= 0; ) {
```
Comment (npmccallum) [inline cmd/list.c]: I suspect the string with options characters here is incorrect.

File: cmd/list.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+#define FMT(t) ((t)->next ? "%s, " : "%s\n")
+#define FMTOP(t) ((t)->next ? "[%s]%s/%s, " : "[%s]%s/%s\n")
+#define FMTP(t, p) (((t)->next || *((p)+1)) ? "%s, " : "%s\n")
+
+    fprintf(stdout, "jwk_type: ");
+    for (type = jose_jwk_types(); type; type = type->next) {
+        fprintf(stdout, FMT(type), type->kty);
+    }
```
Comment (npmccallum) [inline cmd/list.c]: Jose doesn't use braces for single line if/for/etc...

File: Makefile.am (hunk 1 above)
Comment (tiran) [inline Makefile.am]: What do you recommend? lst ? ls ?

File: cmd/list.c (hunk 3 above)
Comment (tiran) [inline cmd/list.c]: Copy and paste ... I'll fix it

File: Makefile.am (hunk 1 above)
Comment (npmccallum) [inline Makefile.am]: dbg? alg? sup?

**Comment by tiran** [comment]
I have moved all hook functions and the static variables to a new file hooks.c. The command is ```cmd/jose sup```

File: cmd/sup.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    for (zipper = jose_jwe_zippers(); zipper; zipper = zipper->next)
+        fprintf(stdout, FMT(zipper), zipper->zip);
+    fprintf(stdout, "\n");
+
+    return EXIT_SUCCESS;
+usage:
+    fprintf(stderr,
+"jose " SUP_USE
+"\n"
+"\nSigns a payload using one or more JWKs and outputs a JWS."
```
Comment (npmccallum) [inline cmd/sup.c]: Copy/paste error.

Comment (tiran) [inline cmd/sup.c]: fixed

File: lib/jose.c [hunk 6]
```diff
@@ -23,6 +23,9 @@
 #include <ctype.h>
 #include <string.h>
 
+#include <dlfcn.h>
+#include <stdio.h>
+
 static json_t *
```
Comment (npmccallum) [inline lib/jose.c]: This seems like it is in the wrong place.

**Comment by npmccallum** [comment]
I just merged a slightly modified version of this patch.

File: lib/jose.c (hunk 6 above)
Comment (npmccallum) [inline lib/jose.c]: I took this out of the final commit, but merged the patch.

**Pull request closed without merging**